### PR TITLE
Application Logic Monitoring 

### PR DESCRIPTION
**Pull request by johanneseschrig**
## Description
Adds application logic monitoring to the prometheus actuator endpoint.

### Counters
Counters are increased in batch steps, to prevent tracing the semantics of a counter to a specific request be comparing time stamps. (current batch size is 5, configurable).
```
# HELP submissionController_requests_total Counts requests to the Submission Controller.
# TYPE submissionController_requests_total counter
submissionController_requests_total{type="all",} 15.0
submissionController_requests_total{type="invalidTan",} 0.0
submissionController_requests_total{type="real",} 10.0
submissionController_requests_total{type="fake",} 5.0
```

### Gauges
Currently, we have a gauge on the delay of the fake requests. 
```
# HELP submissionController_fakeDelay The time that fake requests are delayed to make them indistinguishable from real requests.
# TYPE submissionController_fakeDelay gauge
submissionController_fakeDelay 13.789004980499998
```

### Timers
We time the handling of a submission as a whole, as well as the time spent calling the verification service, and the time spent saving the keys. 

```
# HELP http_server_requests_seconds Time spent handling submission.
# TYPE http_server_requests_seconds summary
http_server_requests_seconds_count{exception="None",method="POST",outcome="SUCCESS",status="200",uri="/version/v1/diagnosis-keys",} 17.0
http_server_requests_seconds_sum{exception="None",method="POST",outcome="SUCCESS",status="200",uri="/version/v1/diagnosis-keys",} 0.545092931
# HELP http_server_requests_seconds_max Time spent handling submission.
# TYPE http_server_requests_seconds_max gauge
http_server_requests_seconds_max{exception="None",method="POST",outcome="SUCCESS",status="200",uri="/version/v1/diagnosis-keys",} 0.27550433
# HELP method_timed_seconds Time spent handling submission.
# TYPE method_timed_seconds summary
method_timed_seconds_count{class="app.coronawarn.server.services.submission.controller.SubmissionController",exception="none",method="submitDiagnosisKey",} 17.0
method_timed_seconds_sum{class="app.coronawarn.server.services.submission.controller.SubmissionController",exception="none",method="submitDiagnosisKey",} 0.058723296
method_timed_seconds_count{class="app.coronawarn.server.services.submission.verification.VerificationServerClient",exception="none",method="verifyTan",} 11.0
method_timed_seconds_sum{class="app.coronawarn.server.services.submission.verification.VerificationServerClient",exception="none",method="verifyTan",} 0.109319729
method_timed_seconds_count{class="app.coronawarn.server.common.persistence.service.DiagnosisKeyService",exception="none",method="saveDiagnosisKeys",} 11.0
method_timed_seconds_sum{class="app.coronawarn.server.common.persistence.service.DiagnosisKeyService",exception="none",method="saveDiagnosisKeys",} 0.088176189
# HELP method_timed_seconds_max Time spent handling submission.
# TYPE method_timed_seconds_max gauge
method_timed_seconds_max{class="app.coronawarn.server.services.submission.controller.SubmissionController",exception="none",method="submitDiagnosisKey",} 0.026234416
method_timed_seconds_max{class="app.coronawarn.server.services.submission.verification.VerificationServerClient",exception="none",method="verifyTan",} 0.060596219
method_timed_seconds_max{class="app.coronawarn.server.common.persistence.service.DiagnosisKeyService",exception="none",method="saveDiagnosisKeys",} 0.067230507
```

### CODE REVIEW

File: services/submission/src/main/java/app/coronawarn/server/services/submission/monitoring/SubmissionControllerMonitor.java [hunk 1]
```diff
@@ -0,0 +1,96 @@
...
+import org.springframework.stereotype.Component;
+
+/**
+ * Provides functionality for monitoring the application logic of the
+ * {@link app.coronawarn.server.services.submission.controller.SubmissionController}.
+ */
+@Component
+@ConfigurationProperties(prefix = "services.submission.monitoring")
+public class SubmissionControllerMonitor {
+
```
Comment (michael-burwig) [inline services/submission/src/main/java/app/coronawarn/server/services/submission/monitoring/SubmissionControllerMonitor.java:37]: .

File: services/submission/src/main/java/app/coronawarn/server/services/submission/monitoring/BatchCounter.java [hunk 2]
```diff
@@ -0,0 +1,61 @@
...
+   */
+  public void increment() {
+    if (batch < batchSize) {
+      batch++;
+    } else {
+      counter.increment(batch);
+      batch = 1.;
+    }
+  }
+
```
Comment (michael-burwig) [inline services/submission/src/main/java/app/coronawarn/server/services/submission/monitoring/BatchCounter.java:60]: .

File: services/submission/src/main/java/app/coronawarn/server/services/submission/monitoring/SubmissionControllerMonitor.java [hunk 3]
```diff
@@ -0,0 +1,96 @@
...
+  }
+
+  public void incrementFake() {
+    fakeRequests.increment();
+  }
+
+  public void incrementInvalidTan() {
+    invalidTanRequests.increment();
+  }
+
```
Comment (michael-burwig) [inline services/submission/src/main/java/app/coronawarn/server/services/submission/monitoring/SubmissionControllerMonitor.java:95]: .

File: services/submission/src/main/java/app/coronawarn/server/services/submission/verification/VerificationServerClient.java [hunk 4]
```diff
@@ -37,6 +38,7 @@
    * @param tan the tan to verify.
    * @return 404 when the tan is not valid.
    */
+  @Timed
```
[the remark this answers is not in the document]
Comment (johanneseschrig) [reply services/submission/src/main/java/app/coronawarn/server/services/submission/verification/VerificationServerClient.java:41]: These get grouped under 
```
# HELP method_timed_seconds Time spent handling submission.
```

so their description would not show up anywhere.

File: services/submission/src/main/java/app/coronawarn/server/services/submission/monitoring/BatchCounter.java [hunk 5]
```diff
@@ -0,0 +1,60 @@
...
+        .description(SUBMISSION_CONTROLLER_REQUESTS_COUNTER_DESCRIPTION)
+        .register(meterRegistry);
+  }
+
+  /**
+   * Increments the {@link BatchCounter}. If the batch size is reached, it is provided to monitoring, else, the internal
+   * counter is incremented.
+   */
+  public void increment() {
+    if (batch < batchSize) {
```
Comment (michael-burwig) [inline services/submission/src/main/java/app/coronawarn/server/services/submission/monitoring/BatchCounter.java:53]: Would prefer batch to be of type ``int`` and cast it to double only when ``counter.increment`` is called.
Either that or make ``batchSize`` a ``Double`` too (weird).

File: services/submission/src/main/java/app/coronawarn/server/services/submission/monitoring/SubmissionControllerMonitor.java [hunk 6]
```diff
@@ -0,0 +1,93 @@
...
+   * @param submissionController the submission controller for which the gauges shall be initialized
+   */
+  public void initializeGauges(SubmissionController submissionController) {
+    Gauge.builder(SUBMISSION_CONTROLLER_CURRENT_FAKE_DELAY, submissionController,
+        __ -> submissionController.getFakeDelay())
+        .description("The time that fake requests are delayed to make them indistinguishable from real requests.")
+        .register(meterRegistry);
+  }
+
+  public void incrementAll() {
```
Comment (michael-burwig) [inline services/submission/src/main/java/app/coronawarn/server/services/submission/monitoring/SubmissionControllerMonitor.java:78]: Please, provide some quick java doc to clarify what is being incremented here or rename the methods to explain it.

File: services/submission/src/main/java/app/coronawarn/server/services/submission/monitoring/BatchCounter.java (hunk 5 above)
Comment (johanneseschrig) [reply services/submission/src/main/java/app/coronawarn/server/services/submission/monitoring/BatchCounter.java:53]: changed

**Pull request merged**